proc/coredump.c

```diff
@@ -0,0 +1,518 @@
+	}
+	if (flags & PROT_EXEC) {
+		phdr.p_flags |= PF_X;
+	}
+
+	phdr.p_align = 0;
+	coredump_encodeChunk(state, (u8 *)&phdr, sizeof(phdr));
+}
+
+
+static void coredump_dumpPhdrs(coredump_threadinfo_t *threadInfo, size_t threadCnt, size_t segCnt, process_t *process, cpu_context_t *ctx, coredump_state_t *state)
+{
+	const size_t THREAD_NOTES_SIZE = sizeof(Elf_Nhdr) +
+			align4(sizeof(PRSTATUS_NAME)) +
+			sizeof(elf_prstatus) +
+			SIZE_COREDUMP_GREGSET +
+			SIZE_COREDUMP_THREADAUX;
+	const size_t NOTES_SIZE = SIZE_COREDUMP_GENAUX + threadCnt * (THREAD_NOTES_SIZE);
+	size_t currentOffset;
+	size_t stackSize;
+	void *userSp;
+	map_entry_t *e;
+	size_t i;
+
+	/* Notes */
+	currentOffset = sizeof(Elf_Ehdr) + sizeof(Elf_Phdr) * (1 + segCnt);
+	coredump_dumpPhdr(PT_NOTE, currentOffset, 0, NOTES_SIZE, 0, state);
+	currentOffset += NOTES_SIZE;
+
+	/* Memory */
+#ifdef COREDUMP_MEM_ALL
+	proc_lockSet(&process->mapp->lock);
+	e = lib_treeof(map_entry_t, linkage, lib_rbMinimum(process->mapp->tree.root));
+	while (e != NULL) {
+		if ((e->prot & PROT_READ) && (e->prot & PROT_WRITE)) {
+			coredump_dumpPhdr(PT_LOAD, currentOffset, e->vaddr, e->size, e->flags, state);
+			currentOffset += e->size;
+		}
+		e = lib_treeof(map_entry_t, linkage, lib_rbNext(&e->linkage));
+	}
+	proc_lockClear(&process->mapp->lock);
+#else
+	userSp = hal_cpuGetUserSP(ctx);
+	stackSize = coredump_stackSize(userSp, process);
+	coredump_dumpPhdr(PT_LOAD, currentOffset, userSp, stackSize, PROT_READ | PROT_WRITE, state);
+	currentOffset += stackSize;
+
+#ifdef COREDUMP_MEM_ALL_STACKS
+	for (i = 1; i < threadCnt; i++) {
+		userSp = hal_cpuGetUserSP(threadInfo[i].userContext);
+		stackSize = coredump_stackSize(userSp, process);
+		coredump_dumpPhdr(PT_LOAD, currentOffset, userSp, stackSize, PROT_READ | PROT_WRITE, state);
+		currentOffset += stackSize;
+	}
+#endif
+
+#endif /* COREDUMP_MEM_ALL */
+}
+
+
+static void coredump_dumpAllMemory(process_t *process, coredump_state_t *state)
+{
+	map_entry_t *e;
+
+	proc_lockSet(&process->mapp->lock);
+	e = lib_treeof(map_entry_t, linkage, lib_rbMinimum(process->mapp->tree.root));
+	while (e != NULL) {
+		if ((e->prot & PROT_READ) && (e->prot & PROT_WRITE)) {
+			coredump_encodeChunk(state, e->vaddr, e->size);
+		}
+		e = lib_treeof(map_entry_t, linkage, lib_rbNext(&e->linkage));
+	}
+	proc_lockClear(&process->mapp->lock);
+}
+
+static size_t coredump_segmentCount(process_t *process)
+{
+	map_entry_t *e;
+	size_t segCnt = 0;
+
+	proc_lockSet(&process->mapp->lock);
+
+	e = lib_treeof(map_entry_t, linkage, lib_rbMinimum(process->mapp->tree.root));
+	while (e != NULL) {
+		if ((e->prot & PROT_READ) && (e->prot & PROT_WRITE)) {
+			segCnt++;
+		}
+		e = lib_treeof(map_entry_t, linkage, lib_rbNext(&e->linkage));
+	}
+
+	proc_lockClear(&process->mapp->lock);
+	return segCnt;
+}
+
+#endif /* COREDUMP */
+
+void coredump_dump(unsigned int n, exc_context_t *ctx)
+{
+#ifndef COREDUMP
+	return;
+#else
+
+	char buff[CORE_BUF_SIZE_MAX];
```

Comment:
You write to this buffer by words/double words, not all targets support possibly unaligned accesses

```suggestion
	char buff[CORE_BUF_SIZE_MAX] __attribute__((aligned(8)));
```